product.go

```diff
@@ -5,26 +5,31 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"github.com/google/go-querystring/query"
 	"strconv"
+
+	"github.com/google/go-querystring/query"
 )
 
 type Product struct {
-	BodyHtml       string      `json:"body_html,omitempty"`
-	CreatedAt      string      `json:"created_at,omitempty"`
-	Handle         string      `json:"handle,omitempty"`
-	ID             int64       `json:"id,omitempty"`
-	Images         interface{} `json:"images,omitempty"`
-	Options        []Option    `json:"options,omitempty"`
-	ProductType    string      `json:"product_type,omitempty"`
-	PublishedAt    string      `json:"published_at,omitempty"`
-	PublishedScope string      `json:"published_scope,omitempty"`
-	Tags           string      `json:"tags,omitempty"`
-	TemplateSuffix string      `json:"template_suffix,omitempty"`
-	Title          string      `json:"title,omitempty"`
-	UpdatedAt      string      `json:"updated_at,omitempty"`
-	Variants       []Variant   `json:"variants,omitempty"`
-	Vendor         string      `json:"vendor,omitempty"`
+	// Any attributes which can be removed from shopify likely should use a pointer type.
+	// This allows us to send empty strings to wipe values out, differentiating between not
+	// wanting to send the value.
+	BodyHTML       *string    `json:"body_html,omitempty"`
+	Handle         *string    `json:"handle,omitempty"`
+	CreatedAt      string     `json:"created_at,omitempty"`
+	ID             int64      `json:"id,omitempty"`
+	Images         *[]Image   `json:"images,omitempty"`
+	Options        []Option   `json:"options,omitempty"`
+	ProductType    *string    `json:"product_type,omitempty"`
+	Published      *bool      `json:"published,omitempty"`
+	PublishedAt    *string    `json:"published_at,omitempty"`
+	PublishedScope *string    `json:"published_scope,omitempty"`
+	Tags           *string    `json:"tags,omitempty"`
+	TemplateSuffix *string    `json:"template_suffix,omitempty"`
+	Title          *string    `json:"title,omitempty"`
+	UpdatedAt      *string    `json:"updated_at,omitempty"`
+	Variants       *[]Variant `json:"variants,omitempty"`
```

Comment:
Interestingly I thought I'd testing the case I want here but it seems to work fine without the pointer.

I'm not worried about the `omitempty` on a `nil`, I'm worried about it on a zero sized slice. I thought I'd tested and found that `make([]string, 0)` would marshal to `null` but it seems like it's actually marshalling to `[]` so I can likely remove these pointers and ensure that the slice is always made, rather than inferring it through `append`

Reply:
https://play.golang.org/p/MrgZkkeCjIv